### PR TITLE
pkg/trace/writer: new queue sender

### PR DESCRIPTION
**Pull request by gbbr**
## Description

This draft PR adds a new writer implementation inside the `pkg/trace/writer/newwriter` package. For now, it only replaces the `TraceWriter` in the trace-agent. The aim is for this writer to fully replace the current one.

## Goals

The motivations and goals behind this were to make all flushing and retries asynchronous, so that further traces can continue to come in and be processed while flushing happens. Previously, flushing the retry queue would result in blocking the entire pipeline, piling up traces in memory and resulting in big memory spikes and OOM kills when any network issues were present (e.g. latency, problems connecting, etc.)

## Benchmark results

A series of small benchmarks were run to reproduce the problem. A [fake server](https://github.com/DataDog/trace-agent-test/tree/master/cmd/fake-edge) was started which replied several times consecutively with 500 responses that took between 2-10 seconds in response time. The exact details of the test (and how `fake-edge` was run) can be seen [here](https://github.com/DataDog/trace-agent-test/blob/master/runs/latency.sh).

#### Before

before

#### After

after

### Blocking

This graph illustrates how long the sampling goroutine was blocked waiting on the send channel, to forward traces to the writer.

#### Before

block-before

#### After

block-after

### Dropped payloads

This graph illustrates how many payloads were dropped as a result of the retry queue getting larger than its allowed size (64MB).

#### Before

dropped-before

#### After

dropped-after

### CODE REVIEW

File: pkg/trace/writer/newwriter/trace.go [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+		}
+	}
+}
+
+func (w *TraceWriter) addSpans(pkg *SampledSpans) {
+	if pkg.Empty() {
+		return
+	}
+
+	atomic.AddInt64(&w.stats.Spans, int64(len(pkg.Trace)+len(pkg.Events)))
```
Comment (clutchski) [inline pkg/trace/writer/newwriter/trace.go]: are these counts overlapping ?

File: pkg/trace/writer/newwriter/trace.go [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+type TraceWriter struct {
+	in       <-chan *SampledSpans
+	hostname string
+	env      string
+	senders  []*sender
+	stop     chan struct{}
+
+	stats *info.TraceWriterInfo
+
+	traces       []*pb.APITrace // traces buffered
```
Comment (clutchski) [inline pkg/trace/writer/newwriter/trace.go]: is this thread-safe between add and flush?

Comment (gbbr) [reply pkg/trace/writer/newwriter/trace.go]: Yes, the `select` statement synchronizes access. The "bigger" changes are in `sender.go`, this file didn't change that much from what we had.

File: pkg/trace/writer/newwriter/trace.go (hunk 1 above)
Comment (gbbr) [reply pkg/trace/writer/newwriter/trace.go]: Not exactly sure what you mean. These are metrics we've always had. Tried to keep the same.

File: pkg/trace/writer/newwriter/sender.go [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+			count:    int(done),
+			duration: time.Since(startTime),
+		})
+	}
+
+	// we reassess the state of the list and check if further flushes need
+	// to be scheduled
+	q.mu.Lock()
+	defer q.mu.Unlock()
+	defer q.wg.Done()
```
Comment (AlexJF) [inline pkg/trace/writer/newwriter/sender.go]: Nit: Can we move this defer to the beginning of `flush` for future-proofing against eventual `return` that may be added above this line at a later point?

File: pkg/trace/writer/newwriter/sender.go [hunk 4]
```diff
@@ -0,0 +1,382 @@
...
+
+	// enqueue the payload
+	q.enqueueLocked(p)
+
+	if !q.scheduled {
+		// no flush is scheduled; start one
+		q.scheduled = true
+		q.wg.Add(1)
+		go q.flush()
+	}
```
Comment (AlexJF) [inline pkg/trace/writer/newwriter/sender.go]: Nit: These 3 things are also done at the end of a `flush`. If we integrated them inside `flush()` itself we could maybe avoid the repetition?

File: pkg/trace/writer/newwriter/sender.go [hunk 5]
```diff
@@ -0,0 +1,382 @@
...
+			stats := &eventData{
+				bytes:    len(p.body),
+				count:    1,
+				duration: time.Since(start),
+				err:      err,
+			}
+			switch err.(type) {
+			case *retriableError:
+				// request failed again, but can be retried
+				q.enqueue(p)
```
Comment (AlexJF) [inline pkg/trace/writer/newwriter/sender.go]: Maybe we should call a `enqueueOrDrop` in the case of a retry with a full queue so we drop the retry and not the more recent data that was since enqueued.

File: pkg/trace/writer/newwriter/sender.go [hunk 6]
```diff
@@ -0,0 +1,382 @@
...
+	if retries > 0 {
+		// some sends failed as retriable; we need to back off a bit
+		q.attempt++
+		q.wg.Add(1)
+		q.timer = time.AfterFunc(backoffDuration(q.attempt), q.flush)
+		return
+	}
+	// all items in the list are new; flush immediately
+	q.wg.Add(1)
+	go q.flush()
```
Comment (AlexJF) [inline pkg/trace/writer/newwriter/sender.go]: Nit: Given this structuring of `flush`, we'll only ever have 1 routine actively on it. Sounds like we might be able to achieve a cleaner implementation if this routine is always the same and simply does something like:

```
for _ := range q.flushReq {
  backoffDuration := flush()
  time.Sleep(backoffDuration)
}
```

With `q.flushReq = make(chan bool, 1)`.

And wherever you are doing `go q.flush()` we could just call:

```
func (q *sender) requestFlush() {
  select {
    case q.flushReq <- true:
       // cool beans
    default:
       // req already there so no point in adding another
}
```

File: pkg/trace/writer/newwriter/sender.go [hunk 7]
```diff
@@ -0,0 +1,382 @@
...
+
+	if q.list.Len() == 0 {
+		// the list is empty; no further flushing needs to be triggered
+		q.attempt = 0
+		q.scheduled = false
+		return
+	}
+	// the list is not empty
+	q.scheduled = true
+	if retries > 0 {
```
Comment (AlexJF) [inline pkg/trace/writer/newwriter/sender.go]: Since this is already treated as a special case (early return), I think it makes more sense to be the first special case to treat instead of `q.list.Len() == 0` (similar to how you usually handle the error scenario close to the place that returns the `err`).

File: pkg/trace/writer/newwriter/sender.go [hunk 8]
```diff
@@ -0,0 +1,382 @@
...
+func (q *sender) recordEvent(t eventType, data *eventData) {
+	if recorder := q.cfg.recorder; recorder != nil {
+		data.host = q.cfg.url.Hostname()
+		recorder.recordEvent(t, data)
+	}
+}
+
+// Flush waits up to 5 seconds for the queue to reach an empty state and for all scheduling
+// to complete before returning.
+func (q *sender) Flush() {
```
Comment (AlexJF) [inline pkg/trace/writer/newwriter/sender.go]: I don't think this should be part of the public interface of a sender since it gives the false impression that payloads are buffered and flushed at arbitrary points in time when in fact flush is attempted after each payload. Also, the fact that no return feedback is given makes it kinda weird as you're never sure if something actually flushed or not.

I think this is mostly being used for tests and for stopping logic. Would it make sense to just call this `Stop`/`Close, wait for empty pipes, free resources and prevent new stuff from being enqueued as you as soon as you called it?

File: pkg/trace/writer/newwriter/sender_test.go [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+	})
+
+	t.Run("accept-many", func(t *testing.T) {
+		assert := assert.New(t)
+		server := newTestServer()
+		defer server.Close()
+
+		s := newSender(testSenderConfig(server.URL))
+		for i := 0; i < climit*4; i++ {
+			s.Push(expectResponses(200))
```
Comment (AlexJF) [inline pkg/trace/writer/newwriter/sender_test.go]: I feel like we could have another test ensuring payloads are processed concurrently according to climit. I don't think any of the existing ones does this.

File: pkg/trace/writer/newwriter/sender.go (hunk 4 above)
Comment (gbbr) [reply pkg/trace/writer/newwriter/sender.go]: I'm not 100% sure I get what you mean here. For the wait group I think it's typical (and makes sense) for it to be outside of the goroutine that it is counting. For the boolean flag, if we integrate it into an asynchronous goroutine, it could cause a race condition in this very function if a new payload comes in before the bool flag is flipped, if that makes sense. That would then potentially cause further goroutines to be created; not a tragedy, but it would be incorrect.

Never the less, I isolated all of this logic out into a new method called `scheduleFlushLocked(delay time.Duration)`. Do you think this is better? It seems nicer to read.

File: pkg/trace/writer/newwriter/sender.go (hunk 3 above)
Comment (gbbr) [reply pkg/trace/writer/newwriter/sender.go]: Yep, that's definitely more correct.

File: pkg/trace/writer/newwriter/sender.go (hunk 7 above)
Comment (gbbr) [reply pkg/trace/writer/newwriter/sender.go]: Oh yeah. That totally looks easier to read. Great suggestion.

File: pkg/trace/writer/newwriter/sender_test.go (hunk 9 above)
Comment (gbbr) [reply pkg/trace/writer/newwriter/sender_test.go]: Yes, it is definitely on my list to add more tests. I just stopped for a bit to get some initial reviews, to make sure I haven't omitted anything in the overall logic, in case someone reveals some drastic changes that will be needed which will in turn require changing more tests.

File: pkg/trace/writer/newwriter/sender.go (hunk 6 above)
Comment (gbbr) [reply pkg/trace/writer/newwriter/sender.go]: The current implementation was a bit inspired from other projects (e.g. Opencensus) and partly motivated by the fact that we have way too many channels and goroutines in the trace-agent. My aim was to avoid using them, unless needed. In a way, it does make sense to have the goroutine you suggest, but from what I've seen in some profiles and given that we already know that the runtime scheduler is one of the next bottlenecks to address in the agent, I didn't want to put more pressure on it.

To clarify, I think that having a goroutine waiting for work could determine the scheduler to "occasionally visit it" to check if it's ready for work. And many times it won't be, so it will be wasted time. This type of "visiting" for nothing already happens quite often. 

At the same time, it is of course debatable that the "always freshly" created goroutines have a similar impact due to calls like `runtime.morestack` etc (which may be optimized in this case because the call is a method; I'm not sure). At the very least, in this scenario, we know for certain that the goroutine which was started has work to do and isn't just idle waiting around.

I would leave it as it is.

File: pkg/trace/writer/newwriter/sender.go (hunk 5 above)
Comment (gbbr) [reply pkg/trace/writer/newwriter/sender.go]: Yep. Great idea. I was pondering around this myself but didn't wanna complicate things. It does make sense. Added it and called it `enqueueAgain`.

File: pkg/trace/writer/newwriter/sender.go (hunk 8 above)
Comment (gbbr) [reply pkg/trace/writer/newwriter/sender.go]: I think this method is honestly the one I've been stuck on the most. I'm not sure what the best approach is. I think that adding more boolean flags throughout of the form `if q.stopped { return }` is not necessary and just makes everything a bit yuckier (I did try it). Furthermore, there is nothing to actually stop.

The logic here was that we try and wait for a bit for the queue to be emptied (potentially doing some more retries) and then return.

I do agree that the name is confusing. But I don't have any good ideas: `stopping`, `waitEmpty`? I'll try `waitEmpty` but leave this comment unresolved for any better ideas.

**Review by gbbr — COMMENTED** [review]
Thanks Alex. I've addressed most of them, and some I had more questions. The ones solved are marked resolved.

File: pkg/trace/writer/newwriter/client.go [hunk 10]
```diff
@@ -43,6 +46,8 @@ func newSenders(cfg *config.AgentConfig, r eventRecorder, path string, climit in
 		Timeout:   10 * time.Second,
 		Transport: transport,
 	}
+	// spread out the the maximum connection limit (climit) between senders
+	maxConns := math.Min(1, float64(climit/len(cfg.Endpoints)))
```
Comment (AlexJF) [inline pkg/trace/writer/newwriter/client.go]: This should be a `math.Max` 😅

File: pkg/trace/writer/newwriter/sender.go [hunk 11]
```diff
@@ -0,0 +1,382 @@
...
+// the sender.
+type eventRecorder interface {
+	// recordEvent notifies that event t has happened, passing details about
+	// the event in data.
+	recordEvent(t eventType, data *eventData)
+}
+
+// maxQueueSize specifies the maximum allowed queue size. If it is surpassed, older
+// items are dropped to make room for new ones.
+var maxQueueSize = 64 * 1024 * 1024 // 64MB; replaced in tests
```
[the remark this answers is not in the document]
Comment (AlexJF) [reply pkg/trace/writer/newwriter/sender.go]: Yeah this won't really be a problem under normal operation. But whenever there's some network outage it directly affects the amount of data that will be lost.

I'm a bit itchy about leaving it hardcoded as 64MB for a particular org might represent 5 minutes of data and for another it might represent 30 seconds. OTOH, we could definitely delay the exposure of this config knob until requested. Making it a fraction of `max_memory` could also work I guess.

This is more of a "product" decision than technical at the end.

File: pkg/trace/writer/newwriter/sender.go (hunk 4 above)
Comment (AlexJF) [reply pkg/trace/writer/newwriter/sender.go]: Yep I much prefer it with this new `scheduleFlushLocked` :+1:

File: pkg/trace/writer/newwriter/sender.go (hunk 8 above)
Comment (AlexJF) [reply pkg/trace/writer/newwriter/sender.go]: My issue was mostly with it being exposed as part of the "public" interface. I'm fine with its current state as `waitEmpty` specially if there's nothing to stop.

File: pkg/trace/writer/newwriter/client.go (hunk 10 above)
Comment (gbbr) [reply pkg/trace/writer/newwriter/client.go]: Haha. Whoops.

File: pkg/trace/writer/newwriter/sender.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (gbbr) [reply pkg/trace/writer/newwriter/sender.go]: Ok, let's leave it open and get more input from product people later whether we want to:

a) Hard code 64MB.
b) Use 12.8% of the `apm_config.max_memory` value (which is 64MB by default).
c) Add an undocumented (?) configuration knob.

My vote stands for b) to avoid having even more knobs and to make it proportional to what the user wants to allocate to the process.

File: pkg/trace/writer/newwriter/sender_test.go (hunk 9 above)
Comment (gbbr) [reply pkg/trace/writer/newwriter/sender_test.go]: I added another commit with a test for this, but I'm not sure it's 100% correct.

File: pkg/trace/writer/newwriter/testserver.go [hunk 12]
```diff
@@ -77,14 +79,32 @@ func (rs *requestStatus) nextResponse() int {
...
+
+// Failed returns the number of connections to which the server responded with a
+// 2xx HTTP status code.
 func (ts *testServer) Accepted() int { return int(atomic.LoadUint64(&ts.accepted)) }
 
 // ServeHTTP responds based on the request body.
 func (ts *testServer) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 	atomic.AddUint64(&ts.total, 1)
+	if v := atomic.AddInt64(&ts.active, 1); v > atomic.LoadInt64(&ts.peak) {
+		atomic.SwapInt64(&ts.peak, v)
```
Comment (AlexJF) [inline pkg/trace/writer/newwriter/testserver.go]: Peak is not guaranteed to be correct with this:
* Routine 1 comes in bumps active to 1, loads peak (0) and is scheduled out.
* Routine 2 comes in bumps active to 2, loads peak (0) and is scheduled out.
* Routine 3 comes in bumps active to 3, loads peak (0) and sets peak to 3.
* Routine 2 comes back in and sets peak to 2.
* Routine 1 comes back in and sets peak to 1.

You need an exclusivity zone surrounding the reading and the writing for this to be correct. Since this is a test server only used for testing I don't think it'd be much of a problem if you added a lock here.

Comment (gbbr) [reply pkg/trace/writer/newwriter/testserver.go]: I'm not a huge expert on the scheduler, but I remember reading somewhere that the scheduler jumps to another routine only when encountering a blocking operation. Nevertheless, I tried what you said and it didn't work much better. I discovered however that it was because the requests were simply server too fast, so I've added an option to create some latency. After doing that, all calls of that test reached the maximum set limit (100). So I think that should do it.

Pushed a new commit with that.

Comment (AlexJF) [reply pkg/trace/writer/newwriter/testserver.go]: I don't usually like to write code based on assumptions of how some piece of the runtime behaves at the current time. Plus, if what you said were true then a routine running an infinite loop would prevent all other routines from running. 

Anyway, I don't think it's much of a matter for this testing code but if this were on production code I'd really prefer the theoretically correct version.

Comment (gbbr) [reply pkg/trace/writer/newwriter/testserver.go]: I 100% agree with that. We shouldn't rely on how the scheduler is assumed to work when writing code, and it should be correct. If you want me to add the lock, happy to do so. I only removed it because it seemed to work without it too, after adding fake latency.

(Side note: I do remember reading some interesting things around `for{}` vs `select{}`, something along the lines: https://stackoverflow.com/questions/18661602/what-does-an-empty-select-do).

File: pkg/trace/writer/newwriter/trace.go (hunk 1 above)
Comment (gbbr) [reply pkg/trace/writer/newwriter/trace.go]: Maybe we should measure events separately 🤔I will add another metric. It seems misleading to include events into "spans"

File: pkg/trace/writer/newwriter/sender.go [hunk 13]
```diff
@@ -0,0 +1,458 @@
...
+				bytes:          p.body.Len(),
+				count:          1,
+				duration:       time.Since(start),
+				err:            err,
+				connectionFill: float64(len(q.climit)) / float64(cap(q.climit)),
+			}
+			switch err.(type) {
+			case *retriableError:
+				// request failed again, but can be retried
+				q.enqueueAgain(p)
```
Comment (djova) [inline pkg/trace/writer/newwriter/sender.go]: Wondering if there's a simpler approach. Could we make this work with only one "pending payloads" channel per sender, and if we need to retry we just throw the payloads back on the same channel? 

If we need to track retries per payload, could wrap it another struct that has a `retries` field that we increment every time it gets retried. 

That way there's only one simple loop reading off of the channel, sending payloads concurrently as they come in (no queueing/flushing), putting them back on the same channel for retries, and adjusting the rate limit according to errors? See my other comment regarding why a rate limit might be a better backoff mechanism than this payload flush strategy.

**Review by djova — CHANGES_REQUESTED** [review]
Left some suggestions, mostly non-blocking.

Main blocking issue is the comment regarding dropping payloads at the sender level.

File: pkg/trace/writer/newwriter/sender.go (hunk 13 above)
Comment (gbbr) [reply pkg/trace/writer/newwriter/sender.go]: > Wondering if there's a simpler approach. Could we make this work with only one "pending payloads" channel per sender, and if we need to retry we just throw the payloads back on the same channel?

Alex also suggested something like this, but my answer stands as part of my other comment of using a channel as a queue.

> If we need to track retries per payload, could wrap it another struct that has a retries field that we increment every time it gets retried.

An error that can be retried says something about the state of the endpoint (e.g. back off a bit) and is not linked specifically to the payload, so it makes no sense to have a per-payload backoff mechanism. It's either that we are backing off from sending to the endpoint, or we are not. It applies in general to all connections.

> That way there's only one simple loop reading off of the channel, sending payloads concurrently as they come in (no queueing/flushing), putting them back on the same channel for retries, and adjusting the rate limit according to errors? See #3688 (comment) regarding why a rate limit might be a better backoff mechanism than this payload flush-interval strategy.

I will spread my answer to this in that very comment and your others.

File: pkg/trace/writer/newwriter/sender.go [hunk 14]
```diff
@@ -0,0 +1,458 @@
...
+
+	// we reassess the state of the list and check if further flushes need
+	// to be scheduled
+	q.mu.Lock()
+	defer q.mu.Unlock()
+	switch {
+	case retries > 0:
+		// some sends failed as retriable; we need to back off a bit
+		q.attempt++
+		delay := backoffDuration(q.attempt)
```
[the remark this answers is not in the document]
Comment (gbbr) [reply pkg/trace/writer/newwriter/sender.go]: > possible danger - if the queue is really backed up due to retries, we may end up doing huge bursts of requests here (up to the connection limit). I don't see any kind of rate limit, so we could potentially send out 64 * 1024 * 1024 payloads all at the same time? On smaller hosts that may cause large CPU or network bursts that could negatively impact other things running on the host.

Yep, this is my least favorite aspect this implementation. The rate limit is currently the connection limit, which limits us to maximum 20 (by default) simultaneous connections. I have spent quite some time thinking about a way around this and explored two ideas:

* Do one request first and see if the endpoint has become available again, if yes, do the rest in parallel. I ruled this one out because it added more complexity as well as prolonged the time it takes to flush by 2x or more, only to cover for some very rarely occurring event. Furthermore, if there is only network latency and no real issues sending to the endpoint, it would likely cause the queue to become full and drop payloads due to the slow flushing. Which brings back the original problem we are trying to solve.

* Use a request which [can be canceled](https://golang.org/pkg/context/#WithCancel). This worked, and when one of the payloads failed, it managed to successfully cancel the rest of the parallel requests. However (this took a long time to figure out) I realized that requests can be canceled also after their bodies reach the server, but there was not yet a response, which means that we could sometimes risk sending the same payload twice. And we don't want that.

I'd say it's fine to live with this. We can explore it later if it really becomes a problem. We might be able to reassess the second point from above and perhaps use the [client trace from the `httptrace` package](https://golang.org/pkg/net/http/httptrace/#ClientTrace) to tell if the request body was written to the server by means of the `WroteRequest` field, in cases where the response didn't make it back because of a cancelation.

But for now, let's keep it simple and not account for extreme edge-cases which result in complicated code. The code is already slightly complicated as is. 

> Would a better backoff strategy be simply to have a rate limit per sender? The rate limit defaults to "no limit", then gets gradually bumped up/down as we see errors (or the lack of errors)?

See my answer [here](https://github.com/DataDog/datadog-agent/pull/3688/files#r294979308).

File: pkg/trace/writer/newwriter/sender.go [hunk 15]
```diff
@@ -0,0 +1,458 @@
...
+	q.size = 0
+	return payloads
+}
+
+// sendPayloads concurrently sends the given list of payloads. It returns
+// the number of payloads that were added back onto the queue to be retried
+// again later due to an error.
+func (q *sender) sendPayloads(payloads []*payload) (done, retries uint64) {
+	var wg sync.WaitGroup
+	for _, p := range payloads {
```
[the remark this answers is not in the document]
Comment (gbbr) [reply pkg/trace/writer/newwriter/sender.go]: > if a payload is ready to be sent, there's no reason not to send it out immediately (unless we're in backoff mode). I see the default flush interval is 0, so in theory we're sending them all out instantly, but still the concept of queueing it up for some flush seems unnecessary.

We _are_ sending them out immediately. There is no waiting. But I see what you mean. This is how the previous implementation worked too. It's absolutely natural to wonder: "why should we queue one payload onto a list, only to drain that list immediately and flush that payload after?". The answer is: code simplicity, readability and maintainability.

Even though it can seem like we are complicating a simple process, the code is now easier to test and reason about. Things happen in this order, and always follow the same code path:

1. Payload comes in
2. Payload gets queued (regardless of what's going on: queue is flushing, queue is not flushing, queue is active or backing off)
    * If a flush isn't scheduled already, one gets scheduled.
4. Queue get flushed at some point, then repeat.

Versus:

1. Payload comes in.
2. We check the state of the queue
    * Are we flushing? Enqueue on one code path.
    * Are we in backoff? Same as above.
    * Are we not flushing? Send on a different code path.

In short, I want to have only one method which flushes, not two: one that flushes immediately, and one that flushes by draining a queue. It just complicates things too much.

Looking at the [previous implementation](https://github.com/DataDog/datadog-agent/blob/a08cd3555c92780bcd6b580799e1e179aaec3c2d/pkg/trace/writer/sender.go), which works a bit like you're hinting at and is a great example of what you're saying, I strongly prefer this one, which even though it can make it seem like we are doing all this for nothing, only to immediately flush, the logic and code path is obvious, simple, easy to reason about, easy to modify/refactor/change and easy to test. 

Looking at resource usage, the cost is probably inexistent, or at the worst completely negligible.

File: pkg/trace/writer/newwriter/sender.go [hunk 16]
```diff
@@ -0,0 +1,458 @@
...
+
+// enqueueLocked adds p onto the queue, making room if needed.
+func (q *sender) enqueueLocked(p *payload) {
+	size := p.body.Len()
+	for q.size+size > maxQueueSize {
+		// make room
+		v := q.list.Remove(q.list.Front())
+		size := v.(*payload).body.Len()
+		q.size -= size
+		q.recordEvent(eventTypeDropped, &eventData{bytes: size, count: 1})
```
[the remark this answers is not in the document]
Comment (gbbr) [reply pkg/trace/writer/newwriter/sender.go]: > Since we already have the watchdog (which refuses payloads when memory gets too high), could we instead rely on that mechanism to drop payloads when we get too backed up here? 

I think we shouldn't always use "watchdog" as a generic motive for having unbounded things. We don't want anything to be unbounded in the application. Watchdog is an "embarrassing" system which should not be needed in well designed program and is something we plan to get rid of sooner or later. I don't think we should have unbounded queues, anywhere. Configurable, maybe, unbounded, no.

> From a user's perspective, the easiest way to think about agent configuration is to say "I will give the agent at most X memory to use".

I completely agree. That's why (as you may have noticed), this implementation has 100% less configuration than the previous one, which had a lot of knobs we've never used. My suggestion (as in a previous comment) was to use 12.8% of `apm_config.max_memory` for this queue (which is by default 64M). We can discuss if the number needs to be bigger, but in my tests, and given past history, it seemed fine. We can always reassess it later on.

I'm also happy to have a dedicated setting for this if we think it's a better idea.

**Review by gbbr — COMMENTED** [review]
@djova thanks for your review. I think your comments are spot on and the questions you've raised were the things I've spent the most time reasoning about. Because they involve some fairly unusual choices, it's good to have discussions based on them.

One (very) important thing to keep in mind, which was possibly the biggest motivation around this rewrite, besides solving the issues which we've learned about through the previous implementation, was to have a new implementation which is easy to test, reason about, refactor and change in the future. I think these are very important qualities.

There are, of course, edge cases and obvious improvements which we could consider (e.g. burst flush) to fix, but in my opinion, the danger of them is so small, that it is not worth the cost that we will pay in code complexity.

I think the point at which we are now is a good place to start and move forward. We can explore with more proficient request cancelation later on if this turns out to be an issue. But thinking about situations which determine retries (timeouts, proxy errors, network down, 5xx responses), I think that in those cases small bursts present no danger to the edge.

File: pkg/trace/writer/newwriter/sender.go (hunk 13 above)
Comment (djova) [reply pkg/trace/writer/newwriter/sender.go]: > An error that can be retried says something about the state of the network (e.g. back off a bit) and is not linked specifically to the payload, so it makes no sense to have a per-payload backoff mechanism.

Good point. Yeah on second thought, I guess no reason to track number of times a payload has been retried due to a retryable error.

File: pkg/trace/writer/newwriter/sender.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (djova) [reply pkg/trace/writer/newwriter/sender.go]: > I think we shouldn't always use "watchdog" as a generic motive for having unbounded things. We don't want anything to be unbounded in the application. Watchdog is an "embarrassing" system which should not be needed in a well designed program and is something we plan to get rid of sooner or later. I don't think we should have unbounded queues, anywhere. Configurable, maybe, unbounded, no.

Good point. I guess "unbounded" in this case is only acceptable assuming watchdog is what we rely on to give it a practical bound. But I agree, watchdog is kind of a hack, so maybe it's best not to rely on that as the primary way to enforce how big the "pending payloads" queue is. 

> I completely agree. That's why (as you may have noticed), this implementation has 100% less configuration than the previous one, which had a lot of knobs we've never used. My suggestion (as in a previous comment) was to use 12.8% of `apm_config.max_memory` for this queue (which is by default 64M). We can discuss if the number needs to be bigger, but in my tests, and given past history, it seemed fine. We can always reassess it later on.
> 
> I'm also happy to have a dedicated setting for this if we think it's a better idea.

Why not have it be effectively 100% of max_memory? If max_memory is being used as a proxy for "this is how much memory I want to devote to pending payloads", then it should be able to be bigger? 

Main concern is that for very high load agents that setting might only be a couple of seconds worth of payloads, so for brief network/service hiccups if it's too small we may end up dropping payloads that we could have saved had the queue been bigger.

File: pkg/trace/writer/newwriter/sender.go [hunk 17]
```diff
@@ -0,0 +1,458 @@
...
+}
+
+// newSender creates a new sender, which uses the given HTTP client, URL and API key
+// to make requests using incoming payloads. The limit of outgoing concurrent connections
+// is defined by climit.
+func newSender(cfg *senderConfig) *sender {
+	return &sender{
+		cfg:    cfg,
+		climit: make(chan struct{}, cfg.maxConns),
+		list:   list.New(),
```
[the remark this answers is not in the document]
Comment (djova) [reply pkg/trace/writer/newwriter/sender.go]: > * "How large should the channel be?"

We're already enforcing a size on the list though, so that would be the channel size? 

> * Draining the channel, as well as removing items from it will result in code that looks like below and would again be very unnatural

Back to my other comments about the whole flushing approach, I'm not sure draining the whole channel is necessary as I don't think flushing a group of pending payloads makes sense since they can't be batched. So all we should be doing is reading them off of the channel one-at-a-time and sending them off asynchronously. 

> Lastly, the previous implementation [also used `list.List` for the retry queue](https://github.com/DataDog/datadog-agent/blob/a08cd3555c92780bcd6b580799e1e179aaec3c2d/pkg/trace/writer/sender.go#L81), which is a natural choice for queues and can grow dynamically without it being pre-allocated. Much easier to reason about.

That probably could have been a channel as well. But as I mentioned already we might not even need a separate retry list/channel as we could just throw the retry payloads back onto the same channel.

File: pkg/trace/writer/newwriter/sender.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (djova) [reply pkg/trace/writer/newwriter/sender.go]: > Yep, this is my least favorite aspect of this implementation. The rate limit is currently the connection limit, which limits us to maximum 20 (by default) simultaneous connections.

Oh ok that's not bad. Yeah we're probably fine then.

> * Use a request which [can be canceled](https://golang.org/pkg/context/#WithCancel).

That's probably not necessary. If we've already sent out a request then can just wait for the timeout to cause it to fail. 

I still think a rate-limiter would be a pretty elegant solution that would simplify a lot of this code. I could work something like this

Main loop in each sender just reads payloads in serially of a channel and sends them off asynchronously:
```
case pkg := <- in:
	rateLimiter.acquire()
	sendAsync(pkg)
```

`rateLimiter.acquire()` blocks the right amount of time to enforce a specified rate limit, configured in something like "payloads per second". There's a [nice implementation of this I've used in Google's guava library](https://guava.dev/releases/19.0/api/docs/index.html?com/google/common/util/concurrent/RateLimiter.html).

`sendAsync` will send the payload asynchronously:
* if retryable error, lower the rate limit according to exponential-backoff logic then throw the payload back on the in-channel. Have reasonable minimum for the rate limit, like one payload/second, so it never gets too long. 
* if success, increase the rate limit

If we are fine with dropping payloads when there's more than the configured maximum of pending payloads, we can just use non-blocking channel write to block when the channel is full, and it'll work the same way you're doing now.

**Review by djova — APPROVED** [review]
Left some more suggestions on how I think this could be made much simpler. No serious problems though so it's ok to proceed if we're all happy with this. 

In general I don't think it's a good idea to be dropping payloads so deep in the pipeline where we have no opportunity to communicate back to the clients via the API that the agent is backed up. Ideally all internal queues are small and have no need to be configurable, then we have only one configurable "pending payloads" queue at the front, and if the agent gets backed up for any reason then that queue fills up and the agent starts rejecting payloads at the API. 

So if we're already investing effort in improving the agent it could be a good idea to move toward that kind of architecture.

File: pkg/trace/writer/newwriter/sender.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (gbbr) [reply pkg/trace/writer/newwriter/sender.go]: Right. This is the same to what @AlexJF suggested in another comment. I tried it out and indeed it seems that everything comes out much easier. I will continue to test that and if it’s all good I’ll close this and open a fresh PR.

**Comment by gbbr** [comment]
> Left some more suggestions on how I think this could be made much simpler. No serious problems though so it's ok to proceed if we're all happy with this.

Thanks for that. Indeed, the channel approach simplifies everything and is much easier to reason about. I have created that on another branch.

> In general I don't think it's a good idea to be dropping payloads so deep in the pipeline where we have no opportunity to communicate back to the clients via the API that the agent is backed up. Ideally all internal queues are small and have no need to be configurable, then we have only one configurable "pending payloads" queue at the front, and if the agent gets backed up for any reason then that queue fills up and the agent starts rejecting payloads at the API.

I agree with you. And that's the direction we should be heading to. But I think we are asking a bit too much from this PR now. We will address this issue later on when we work on the API too, as a whole. One step at a time.

I am also adding (undocumented) configuration knobs and defaults based on other settings for the queue size and connection limit.

I am closing this PR and opening a fresh one.
